compiler/rustc_hir_typeck/src/delegation.rs

```diff
@@ -0,0 +1,372 @@
+        _ => FnKind::Free,
+    }
+}
+
+struct GenericDefsCollector<'tcx> {
+    tcx: TyCtxt<'tcx>,
+    info: GenericDefsMap<'tcx>,
+    def_id: LocalDefId,
+}
+
+impl<'tcx> GenericDefsCollector<'tcx> {
+    fn new(tcx: TyCtxt<'tcx>, def_id: LocalDefId) -> Self {
+        Self { tcx, info: GenericDefsMap::default(), def_id }
+    }
+
+    fn collect(tcx: TyCtxt<'tcx>, def_id: LocalDefId) -> GenericDefsMap<'tcx> {
+        let mut collector = GenericDefsCollector::new(tcx, def_id);
+        let caller_kind = fn_kind(tcx, def_id.into());
+        // generics are not yet supported for associated delegation items
+        if caller_kind == FnKind::Free {
+            let path = collector.check_call();
+            path.visit_with(&mut collector);
+        }
+
+        collector.info
+    }
+
+    // Collect generic parameters definitions during callee type traversal according to
+    // encountered inference variables.
+    fn extract_info_from_def(&mut self, def_id: DefId, args: GenericArgsRef<'tcx>) {
+        let generics = self.tcx.generics_of(def_id);
+
+        for (idx, arg) in args.iter().enumerate() {
+            if let Some(ty) = arg.as_type()
+                && let ty::Infer(ty::InferTy::TyVar(ty_vid)) = ty.kind()
+            {
+                self.info.ty_defs.insert(*ty_vid, generics.param_at(idx, self.tcx).clone());
+            } else if let Some(re) = arg.as_region()
+                && let ty::RegionKind::ReVar(inf_re) = re.kind()
+            {
+                self.info.region_defs.insert(inf_re, generics.param_at(idx, self.tcx).clone());
+            }
+        }
+
+        self.info.defs.insert(def_id, args);
+    }
+
+    // Extract callee type from the call path. Should only be used for
+    // non-associated delegation items. For example in
+    //
+    // trait Trait<T> {
+    //     fn foo<U>(&self, x: U, y: T);
+    // }
+    //
+    // reuse <u16 as Trait<_>>::foo;
+    //
+    // it will return `FnDef(DefId(Trait::foo), [u16, ?0t, ?1t])`.
+    fn check_call(&self) -> Ty<'tcx> {
+        let body = self.tcx.hir().body_owned_by(self.def_id);
+        let body = self.tcx.hir().body(body);
+
+        let (expr, callee, args) = match body.value.kind {
+            hir::ExprKind::Block(
+                hir::Block {
+                    expr: expr @ Some(hir::Expr { kind: hir::ExprKind::Call(callee, args), .. }),
+                    ..
+                },
+                _,
+            ) => (expr.unwrap(), callee, args),
+            _ => unreachable!(),
+        };
+
+        let infcx = self.tcx.infer_ctxt().ignoring_regions().build();
+        // FIXME: cycle error on `with_opaque_type_inference`
+        let root_ctxt =
+            TypeckRootCtxt::new_with_infcx(self.tcx, self.def_id, infcx);
+        let param_env = ty::ParamEnv::empty();
+        let fcx = FnCtxt::new(&root_ctxt, param_env, self.def_id);
+        fcx.check_expr_with_expectation_and_args(
```

Comment:
I don't think we should ever type check bodies to produce signatures, only signatures should be analyzed to produce signatures.

(Well, until we start implementing [step 3](https://github.com/rust-lang/rfcs/blob/923b8e65639de80a93ef2b9770ba1c976c24cf5d/text/0000-fn-delegation.md#implementation-steps), but I'm not sure we even need that part anymore.)

Reply:
To support generic arguments in delegation item we should somehow extract type information from path. Type check is required for this. And even if we don't support explicit inference variables, it's convenient to use an implicit ones, produced by the type check, to find generic parameter definitions.